hnn_core/network_builder.py

```diff
@@ -105,6 +107,10 @@ def simulation_time():
         for spike_vec in spikegids_list:
             neuron_net._all_spikegids.append(spike_vec)
 
+        for node_vsoma in vsoma_list:
+            vsoma_data = {k: v.to_python() for k, v in node_vsoma.items()}
+            neuron_net._all_vsoma.update(vsoma_data)
```

Comment:
I'm still trying to wrap my head around how the various nrn and PC calls work, but it looks like `vsoma_list` contains hoc objects buried in it. Is there anyway to simplify `vsoma_list` or `neuron_net._vsoma` so that these variables are more transparent? Maybe `NetworkBuilder().vsoma` could be assigned to a special nrn hoc object that allows `_PC.py_gather()` to instantiate a python-only object?

Reply:
That is correct, but you'll notice that in the immediately preceding lines that `spiketimes_list` and `spikegids_list` also contain hoc objects. The only difference is they are converted to python objects in the function:
```py
    def get_data_from_neuron(self):
        """Get copies of spike data that are pickleable"""

        from copy import deepcopy
        data = (self._all_spiketimes.to_python(),
                self._all_spikegids.to_python(),
                deepcopy(self.net.gid_dict),
                deepcopy(self._all_vsoma))
        return data
```
The main difference is where the python conversion is made. Perhaps as we restructure the spikes class we can change the variable name?

In either case I'll likely need to restructure this as I think using a dictionary is breaking MPI runs (but not joblib) for some reason. I like the suggestion of keeping `NetworkBuilder().vsoma` as a hoc object, perhaps that will help with the MPI compatability...

Reply:
> That is correct, but you'll notice that in the immediately preceding lines that `spiketimes_list` and `spikegids_list` also contain hoc objects.

Oh I see. I had thought that `_PC.py_gather()` was concatenating hoc objects and spitting out a single python object, but I see now that that is not the case.

> In either case I'll likely need to restructure this as I think using a dictionary is breaking MPI runs (but not joblib) for some reason.

I'm wondering if your issue with MPI arrises because you extract the hoc-converted-to-python data too soon.

Reply:
I didn't read carefully but I have an intuition that dictionary MPI because it's not ordered. The MPI calls depend on having things called in the right order. There is a node-specific list called `gid_list` that you should use when needed to make MPI happy.

Reply:
also you might want to rebase before debugging further

Reply:
unfortunately debugging MPI is notoriously hard. You will have to insert print statements and check where the code is failing ... also trying printing what is in `stderr`. Hope these hints hellp